File: Bit_Manipulation/Count_Bits_Flip_A_B.py
# Count number of bits needed to be flipped to convert A to B


def count_bits_flip(a, b):
    # XOR a and b to get 1 on opposite value bit position
    c = a ^ b

    # initialise the counter for 1
    count = 0

    # count the number of 1s while there is 1 in a ^ b
    while c != 0:
        count += 1
        c &= (c-1)

    # return the count of 1s
    return count


# 2 = 0010
# 8 = 1000
print(count_bits_flip(2, 8))

File: Bit_Manipulation/Count_Bits_Flip_A_B.cpp
/* 
Count number of bits needed to be flipped to convert from A to B.
*/ 

#include <iostream>

using namespace std;

int count_bits_flip(int a, int b){
    // Perform a xor b
    int c = a ^ b;

    int count = 0;

    // count number of 1s in c
    while (c != 0){
        c &= (c-1);
        count += 1;
    }

    return count;
}

int main(){
    int a = 15; // 15 = (1111)
    int b = 2;  // 2  = (0010)
    cout << "Number of bits needed to be flipped to convert from " << a << 
    " to " << b << " = " << count_bits_flip(a, b) << endl;
    return 0;
}
